ch_backup/clickhouse/disks.py

```diff
@@ -149,16 +150,53 @@ def _create_temporary_disk(
         self._created_disks[tmp_disk_name] = source_disk
         self._disks[tmp_disk_name] = disk_config
 
-    def copy_part(
+    def copy_parts(
+        self,
+        backup_meta: BackupMetadata,
+        parts_to_copy: List[Tuple[Table, PartMetadata]],
+        max_proccesses_count: int,
+        keep_going: bool,
+    ) -> None:
+        """
+        Copy parts from temporary cloud storage disk to actual.
+
+        If clickhouse greater or equal than 24.1 then we are able to use s3-server-side copy.
+        Spawns no more than max_processes_count of clickhouse-disks subproceses to copy part from tmp disk.
+        """
+
+        with ThreadPoolExecutor(max_workers=max_proccesses_count) as executor:
+            # Can't use map function here. The map method returns a generator
+            # and it is not possible to resume a generator after an exception occurs.
+            # https://peps.python.org/pep-0255/#specification-generators-and-exception-propagation
+            futures_to_part = {
+                executor.submit(
+                    self._run_copy_command, backup_meta, part[0], part[1]
+                ): part[1].name
+                for part in parts_to_copy
+            }
+            for future in futures_to_part:
```

Comment:
Suggest using `as_completed` to detect exceptions early https://github.com/yandex/ch-backup/blob/985571a3ae1c5cdf5f7b15b24679e2d2a3e79556/ch_backup/storage/async_pipeline/base_pipeline/exec_pool.py#L65